### PR TITLE
Core chain locks v5

### PR DESCRIPTION
**Pull request by shumkov**
## Description

_Please add a description of the changes that this PR introduces and the files that
are the most critical to review._ 

Closes: #XXX

### CODE REVIEW

File: abci/example/counter/counter.go [hunk 1]
```diff
@@ -107,11 +109,11 @@ func (app *Application) Query(reqQuery types.RequestQuery) types.ResponseQuery {
...
 func (app *Application) EndBlock(reqEndBlock types.RequestEndBlock) types.ResponseEndBlock {
-	if app.HasChainLocks {
-		app.CurrentChainLockHeight = uint32(int32(app.CurrentChainLockHeight) + app.ChainLockStep)
-		chainLock := types1.NewMockChainLock(app.CurrentChainLockHeight)
+	if app.HasCoreChainLocks {
+		app.CurrentCoreChainLockHeight = uint32(int32(app.CurrentCoreChainLockHeight) + app.CoreChainLockStep)
+		chainLock := types1.NewMockChainLock(app.CurrentCoreChainLockHeight)
 		return types.ResponseEndBlock{
-			NextChainLockUpdate: chainLock.ToProto(),
+			CoreChainLockUpdate: chainLock.ToProto(),
```
Comment (QuantumExplorer) [inline abci/example/counter/counter.go]: I would like to keep the Next part here to differentiate a chain lock update and this. Which is not an update in consensus, but an update to the next chain lock that the node would propose.

File: blockchain/msgs_test.go [hunk 2]
```diff
@@ -82,7 +82,7 @@ func TestBcStatusResponseMessageValidateBasic(t *testing.T) {
 func TestBlockchainMessageVectors(t *testing.T) {
 	chainLock := types.NewMockChainLock(1)
 
-	block := types.MakeBlock(int64(3),chainLock.CoreBlockHeight,&chainLock,[]types.Tx{types.Tx("Hello World")}, nil, nil)
+	block := types.MakeBlock(int64(3),chainLock.BlockHeight,&chainLock,[]types.Tx{types.Tx("Hello World")}, nil, nil)
```
Comment (QuantumExplorer) [inline blockchain/msgs_test.go]: This is very confusing. As this could mean the tendermint BlockHeight at which the chain lock was introduced to platform.

File: crypto/bls12381/bls12381.go [hunk 3]
```diff
@@ -54,9 +54,6 @@ func (privKey PrivKey) Bytes() []byte {
 // If these conditions aren't met, Sign will panic or produce an
 // incorrect signature.
 func (privKey PrivKey) Sign(msg []byte) ([]byte, error) {
-	if len(privKey.Bytes()) != PrivateKeySize {
```
Comment (QuantumExplorer) [inline crypto/bls12381/bls12381.go:57]: Why did you remove this check?

File: crypto/bls12381/bls12381.go [hunk 4]
```diff
@@ -70,9 +67,6 @@ func (privKey PrivKey) Sign(msg []byte) ([]byte, error) {
 //
 // Panics if the private key is not initialized.
 func (privKey PrivKey) PubKey() crypto.PubKey {
-	if len(privKey.Bytes()) != PrivateKeySize {
```
Comment (QuantumExplorer) [inline crypto/bls12381/bls12381.go:73]: Why did you remove this check?

File: proto/tendermint/state/types.proto [hunk 5]
```diff
@@ -42,37 +42,37 @@ message State {
...
+  int64                    last_block_height = 3;
+  tendermint.types.BlockID last_block_id     = 4
       [(gogoproto.nullable) = false, (gogoproto.customname) = "LastBlockID"];
-  google.protobuf.Timestamp last_block_time = 6
+  google.protobuf.Timestamp last_block_time = 5
       [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
 
-  tendermint.types.ChainLock last_chain_lock = 7 [(gogoproto.nullable) = false];
-  tendermint.types.ChainLock next_chain_lock = 8 [(gogoproto.nullable) = false];
+  tendermint.types.CoreChainLock last_core_chain_lock = 100 [(gogoproto.nullable) = false];
```
Comment (QuantumExplorer) [inline proto/tendermint/state/types.proto:54]: This will lead to our system not being backward compatible if any value ever comes before 100. This should not be done.

File: proto/tendermint/types/block.proto [hunk 6]
```diff
@@ -8,9 +8,9 @@ import "tendermint/types/types.proto";
...
 
 message Block {
-  Header                        header      = 1 [(gogoproto.nullable) = false];
-  Data                          data        = 2 [(gogoproto.nullable) = false];
-  ChainLock                     chain_lock  = 3 [(gogoproto.nullable) = true];
-  tendermint.types.EvidenceData evidence    = 4 [(gogoproto.nullable) = false];
-  Commit                        last_commit = 5;
+  Header                        header           = 1   [(gogoproto.nullable) = false];
+  Data                          data             = 2   [(gogoproto.nullable) = false];
+  CoreChainLock                 core_chain_lock  = 100 [(gogoproto.nullable) = true];
```
Comment (QuantumExplorer) [inline proto/tendermint/types/block.proto:13]: This should not be set to 100

File: proto/tendermint/types/types.proto [hunk 7]
```diff
@@ -57,36 +57,36 @@ message BlockID {
...
   // basic block info
-  tendermint.version.Consensus version                  = 1 [(gogoproto.nullable) = false];
-  string                       chain_id                 = 2 [(gogoproto.customname) = "ChainID"];
-  int64                        height                   = 3;
-  uint32                       core_chain_locked_height = 4;
-  google.protobuf.Timestamp    time                     = 5 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
+  tendermint.version.Consensus version  = 1 [(gogoproto.nullable) = false];
+  string                       chain_id = 2 [(gogoproto.customname) = "ChainID"];
+  int64                        height   = 3;
+  uint32                       core_chain_locked_height = 100;
```
Comment (QuantumExplorer) [inline proto/tendermint/types/types.proto:63]: Same here.

File: abci/example/counter/counter.go (hunk 1 above)
Comment (shumkov) [reply abci/example/counter/counter.go]: Yeah, but it's ABCI response which is just transport interface. I was thinking just to keep the name consistent with `ValidatorUpdates` and `ConsensusParamUpdates`.

File: blockchain/msgs_test.go (hunk 2 above)
Comment (shumkov) [reply blockchain/msgs_test.go]: Renamed everywhere to coreChainLock. It kinda lools strange `coreChainLock.CoreBlockHeight`. It's already obvious that it's core height by context. Usually, we don't name `block.blockId`, `block.blockHeight` coz it
s obvious by instance/struct name.

File: crypto/bls12381/bls12381.go (hunk 4 above)
Comment (shumkov) [reply crypto/bls12381/bls12381.go:73]: Oh.. my bad.

File: blockchain/msgs_test.go (hunk 2 above)
Comment (QuantumExplorer) [reply blockchain/msgs_test.go]: actually blocks do have a blockId

Comment (shumkov) [reply blockchain/msgs_test.go]: Hehe.. It was a bad example. You caught me.

**Comment by QuantumExplorer** [comment]
I will merge this in for now. However I am going to change CoreChainLockUpdate to NextCoreChainLockUpdate and reenable the two checks that were removed in error.
